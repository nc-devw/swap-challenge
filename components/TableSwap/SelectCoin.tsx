import React from "react";
import Image from "next/image";
import { CoinsBalance } from ".";
import ExpandMore from "@/public/assets/icons/ExpandMore";

const SelectCoin = ({
  coins,
  currency,
  setCurrency,
  value,
  setValue,
  swapCurrency,
  setSwapValue,
  name,
}: {
  coins: CoinsBalance[];
  currency: CoinsBalance;
  setCurrency: React.Dispatch<React.SetStateAction<CoinsBalance>>;
  value: string;
  setValue: React.Dispatch<React.SetStateAction<string>>;
  swapCurrency: CoinsBalance;
  setSwapValue: React.Dispatch<React.SetStateAction<string>>;
  name?: string;
}) => {
  const [open, setOpen] = React.useState(false);

  const handleCurrency = (coin: CoinsBalance) => {
    setCurrency(coin);
    setOpen(false);
  };

  const handleChange = (
    ev: React.FormEvent<HTMLInputElement>,
    maxValue: number
  ) => {
    const currentValue = ev.currentTarget.value;

    if (currentValue === "") {
      setValue(currentValue);
    }
    if (!/^\d+\.{0,1}\d{0,15}$/.test(currentValue)) {
      return;
    }

    let newValue = "";

    if (Number(currentValue) > maxValue && name !== "output") {
      newValue = maxValue.toString();
    } else {
      newValue = ev.currentTarget.value;
    }

    const swapValue =
      (Number(newValue) * currency.usdPrice) / swapCurrency.usdPrice;
    setSwapValue(swapValue.toString());
    return setValue(newValue);
  };

  return (
    <div className="flex flex-col items-center">
      <div className="flex relative">
        <div
          className="flex cursor-pointer"
          onClick={() => {
            setOpen((prevState) => !prevState);
          }}
        >
          <Image
            alt="cripto"
            src={currency?.iconUrl || "/"}
            width={20}
            height={20}
          />
          <span className="ml-2 uppercase font-bold">{currency?.symbol}</span>
          <ExpandMore className="h-6 w-6" />
        </div>
        <div className="pl-6 text-sm text-gray-400">
          <span>
            Balance: {parseFloat(currency?.currentQuantity.toFixed(5))}
          </span>
        </div>
        <div
          style={{ display: open ? "block" : "none" }}
          className="shadow rounded absolute bg-white z-10 left-20 top-2 w-24 border"
        >
          {coins.map(
            (coin: CoinsBalance) =>
              coin.symbol !== currency?.symbol && (
                <div
                  className="flex items-center p-2 m-2 hover:text-slate-50 hover:bg-violet-500 cursor-pointer rounded"
                  key={coin.symbol}
                  onClick={() => handleCurrency(coin)}
                >
                  <Image
                    alt="cripto"
                    src={coin.iconUrl}
                    width={20}
                    height={20}
                  />
                  <span className="ml-2 font-bold uppercase text-sm">
                    {coin.symbol}
                  </span>
                </div>
              )
          )}
        </div>
      </div>
      <div className="m-2">
        <input
          className="bg-gray-100 rounded p-2"
          type="text"
          value={value}
          onChange={(ev) => handleChange(ev, currency?.currentQuantity)}
        />
      </div>
    </div>
  );
};

export default SelectCoin;
