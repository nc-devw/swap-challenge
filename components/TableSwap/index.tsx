import { AxiosResponse } from "axios";
import React from "react";
import { useAppContext } from "../../context/global";
import useRestClient from "../../hooks/useRestClient";
import { Coin } from "../../services/transforms/coins";
import Card from "../Card";

type Props = {};

export interface CoinsBalance extends Coin {
  currentQuantity: number;
  currentUsdPrice?: number;
}

const TableSwap = (props: Props) => {
  const { response }: { response: AxiosResponse<Coin[], any> | undefined } =
    useRestClient({
      method: "GET",
      url: `/coins`,
    });

  const { user } = useAppContext();

  const [balance, setBalance] = React.useState(0);
  const [coins, setCoins] = React.useState<CoinsBalance[]>([]);

  React.useEffect(() => {
    if (
      response?.data.length === 0 ||
      user?.coins.length === 0 ||
      coins.length !== 0
    ) {
      return;
    }

    response?.data.forEach((coinData) => {
      const coin = user?.coins?.find((coin) => coin.symbol === coinData.symbol);

      if (!coin) {
        let newArray = [...coins, { ...coinData, currentQuantity: 0 }];
        setCoins(newArray);
        return;
      }
      const currentUsdPrice = coin.quantity * coinData.usdPrice;
      setBalance((prevState) => prevState + currentUsdPrice);

      setCoins((prevState) => [
        ...prevState,
        {
          ...coinData,
          currentQuantity: coin.quantity,
          currentUsdPrice,
        },
      ]);

      return;
    });

    return;
  }, [response, user]);
  return (
    <Card>
      <div className="flex flex-col">
        <div className="text-center p-4 border-b">
          <p className="uppercase font-bold text-lg">Swap</p>
          <p className="font-bold text-small text-gray-400">
            Trade tokens in an instant
          </p>
        </div>
        <div className="flex flex-col border-b p-4">
          <div className="flex justify-between">
            <div className="flex">
              {/* // select */}
              <span>icon</span>
              <span>BNB</span>
              <span>chevron</span>
            </div>
            <div>
              <span>Balance: 0.0017</span>
            </div>
          </div>
          <div>
            <span>chevron de swap</span>
          </div>
          <div className="flex justify-between">
            <div className="flex">
              {/* // select */}
              <span>icon</span>
              <span>BTC</span>
              <span>chevron</span>
            </div>
            <div>
              <span>Balance: 0.0017</span>
            </div>
          </div>
        </div>
        <div className="text-center p-4">
          <button className="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded">
            Swap
          </button>
        </div>
      </div>
    </Card>
  );
};

export default TableSwap;
