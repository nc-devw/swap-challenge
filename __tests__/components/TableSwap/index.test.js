import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import TableSwap from "@/components/TableSwap";

const mockCoins = [
  {
    name: "Bitcoin",
    symbol: "btc",
    iconUrl:
      "https://assets.coingecko.com/coins/images/1/small/bitcoin.png?1547033579",
    usdPrice: 31661,
  },
  {
    name: "Ethereum",
    symbol: "eth",
    iconUrl:
      "https://assets.coingecko.com/coins/images/279/small/ethereum.png?1595348880",
    usdPrice: 1994.29,
  },
  {
    name: "USD Coin",
    symbol: "usdc",
    iconUrl:
      "https://assets.coingecko.com/coins/images/6319/small/USD_Coin_icon.png?1547042389",
    usdPrice: 1,
  },
];

jest.mock("@/context/global", () => ({
  useAppContext: () => ({
    assets: [
      {
        symbol: "btc",
        quantity: 0.002,
      },
      {
        symbol: "usdc",
        quantity: 100,
      },
    ],
  }),
}));

describe("Table Swap", () => {
  it("should render correctly", () => {
    const { asFragment } = render(<TableSwap coins={mockCoins} />);
    expect(asFragment()).toBeDefined();
    expect(asFragment()).toMatchSnapshot();
  });
});
