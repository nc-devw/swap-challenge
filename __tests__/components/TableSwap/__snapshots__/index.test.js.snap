// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Swap should render correctly 1`] = `
<DocumentFragment>
  <div
    class="rounded-lg m-12 border shadow bg-white w-80"
  >
    <div
      class="flex flex-col"
    >
      <div
        class="text-center p-4 border-b"
      >
        <p
          class="uppercase font-bold text-lg"
        >
          Swap
        </p>
        <p
          class="font-bold text-small text-gray-400"
        >
          Trade tokens in an instant
        </p>
      </div>
      <div
        class="flex flex-col border-b p-4 items-center"
      >
        <div
          class="flex"
        >
          <div
            class="flex"
          >
            <div
              class="flex flex-col items-center"
            >
              <div
                class="flex relative"
              >
                <div
                  class="flex cursor-pointer"
                >
                  <span
                    style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
                  >
                    <span
                      style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
                    >
                      <img
                        alt=""
                        aria-hidden="true"
                        src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2720%27%20height=%2720%27/%3e"
                        style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                      />
                    </span>
                    <img
                      alt="cripto"
                      data-nimg="intrinsic"
                      decoding="async"
                      src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                      style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                    />
                    <noscript />
                  </span>
                  <span
                    class="ml-2 uppercase font-bold"
                  >
                    eth
                  </span>
                  <svg
                    class="h-6 w-6"
                    viewBox="0 0 50 50"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M24 30.75 12 18.75 14.15 16.6 24 26.5 33.85 16.65 36 18.8Z"
                    />
                  </svg>
                </div>
                <div
                  class="pl-6 text-sm text-gray-400"
                >
                  <span>
                    Balance: 0
                  </span>
                </div>
                <div
                  class="shadow rounded absolute bg-white z-10 left-20 top-2 w-24 border"
                  style="display: none;"
                >
                  <div
                    class="flex items-center p-2 m-2 hover:text-slate-50 hover:bg-violet-500 cursor-pointer rounded"
                  >
                    <span
                      style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
                    >
                      <span
                        style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
                      >
                        <img
                          alt=""
                          aria-hidden="true"
                          src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2720%27%20height=%2720%27/%3e"
                          style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                        />
                      </span>
                      <img
                        alt="cripto"
                        data-nimg="intrinsic"
                        decoding="async"
                        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                      />
                      <noscript />
                    </span>
                    <span
                      class="ml-2 font-bold uppercase text-sm"
                    >
                      usdc
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="m-2"
              >
                <input
                  class="bg-gray-100 rounded p-2"
                  type="text"
                  value="0"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="pt-2 pb-3"
        >
          <svg
            class="w-8 h-8 fill-violet-500 hover:fill-violet-600 cursor-pointer"
            viewBox="0 0 50 50"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M28.8 38.2V23.9Q28.8 23.25 29.225 22.825Q29.65 22.4 30.3 22.4Q30.95 22.4 31.375 22.825Q31.8 23.25 31.8 23.9V38.25L36.9 33.2Q37.3 32.8 37.9 32.8Q38.5 32.8 38.95 33.25Q39.4 33.7 39.4 34.3Q39.4 34.9 38.95 35.35L31.4 42.9Q31.15 43.15 30.875 43.25Q30.6 43.35 30.35 43.35Q30.05 43.35 29.775 43.25Q29.5 43.15 29.3 42.9L21.7 35.25Q21.3 34.85 21.3 34.25Q21.3 33.65 21.75 33.2Q22.15 32.75 22.775 32.775Q23.4 32.8 23.85 33.25ZM16.1 9.7 11.1 14.7Q10.7 15.1 10.1 15.1Q9.5 15.1 9.05 14.65Q8.6 14.2 8.6 13.6Q8.6 13 9.05 12.55L16.6 5Q16.85 4.75 17.1 4.65Q17.35 4.55 17.65 4.55Q17.95 4.55 18.2 4.65Q18.45 4.75 18.7 5L26.3 12.6Q26.7 13 26.7 13.6Q26.7 14.2 26.25 14.65Q25.8 15.1 25.2 15.1Q24.6 15.1 24.15 14.65L19.1 9.65V24Q19.1 24.65 18.675 25.075Q18.25 25.5 17.6 25.5Q16.95 25.5 16.525 25.075Q16.1 24.65 16.1 24Z"
            />
          </svg>
        </div>
        <div
          class="flex"
        >
          <div
            class="flex"
          >
            <div
              class="flex flex-col items-center"
            >
              <div
                class="flex relative"
              >
                <div
                  class="flex cursor-pointer"
                >
                  <span
                    style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
                  >
                    <span
                      style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
                    >
                      <img
                        alt=""
                        aria-hidden="true"
                        src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2720%27%20height=%2720%27/%3e"
                        style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                      />
                    </span>
                    <img
                      alt="cripto"
                      data-nimg="intrinsic"
                      decoding="async"
                      src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                      style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                    />
                    <noscript />
                  </span>
                  <span
                    class="ml-2 uppercase font-bold"
                  >
                    usdc
                  </span>
                  <svg
                    class="h-6 w-6"
                    viewBox="0 0 50 50"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M24 30.75 12 18.75 14.15 16.6 24 26.5 33.85 16.65 36 18.8Z"
                    />
                  </svg>
                </div>
                <div
                  class="pl-6 text-sm text-gray-400"
                >
                  <span>
                    Balance: 100
                  </span>
                </div>
                <div
                  class="shadow rounded absolute bg-white z-10 left-20 top-2 w-24 border"
                  style="display: none;"
                >
                  <div
                    class="flex items-center p-2 m-2 hover:text-slate-50 hover:bg-violet-500 cursor-pointer rounded"
                  >
                    <span
                      style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
                    >
                      <span
                        style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
                      >
                        <img
                          alt=""
                          aria-hidden="true"
                          src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2720%27%20height=%2720%27/%3e"
                          style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                        />
                      </span>
                      <img
                        alt="cripto"
                        data-nimg="intrinsic"
                        decoding="async"
                        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                      />
                      <noscript />
                    </span>
                    <span
                      class="ml-2 font-bold uppercase text-sm"
                    >
                      eth
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="m-2"
              >
                <input
                  class="bg-gray-100 rounded p-2"
                  type="text"
                  value="0"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="text-center p-4"
      >
        <button
          class="text-white font-bold py-2 px-4 rounded cursor-not-allowed bg-gray-400 hover:bg-gray-600"
        >
          Swap
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
